### PR TITLE
[msbuild/tests] Ensure our FrameworkList files have correct assemblies

### PR DESCRIPTION
**Pull request by VincentDondain**
One mono bump (https://github.com/xamarin/xamarin-macios/commit/6f2ebedb74fe8da250c8f53b61ffae2eb86fc711#diff-e801bb766cbaad95b50b1487b865f971) changed our `Xamarin.iOS-FrameworkList.xml.in` (and the tvOS and watchOS ones) and added 2 files

```
<File AssemblyName="System.Buffers" Version="4.0.99.0" />
<File AssemblyName="System.Memory" Version="4.0.99.0" />
```

Around the same time this change was made in the IDE: https://github.com/mono/monodevelop/blob/e355f65870da4847d77ed1f3f2d07d1f78889cec/main/src/core/MonoDevelop.Core/MonoDevelop.Core.Assemblies/TargetFramework.cs#L260
That means that VSMac, when creating a new cache file listing the assemblies, won’t scan the assemblies in those directories `/Library/Frameworks/Xamarin.iOS.framework/Versions/Current/lib/mono/Xamarin.iOS`.
This resulted in an incomplete cache file missing most assemblies leading to the IDE showing red checkmarks for known and installed assemblies.

Ideally the FrameworkList files we ship should have a list of all the assemblies with PublicKeyToken and friends (the framework shouldn't be lazy and expect the IDEs to generate that file for us).

This commit ships a test using the same logic as the IDE to populate the FrameworkList file. The test will report any missing assembly and print the xml element that should be added to the file.
E.g: `<File AssemblyName="FSharp.Core" Version="3.98.4.0" PublicKeyToken="b03f5f7f11d50a3a" ProcessorArchitecture="MSIL" />`
Then it's only a matter of copy/pasting that (:

Note: even though this is the msbuild tests for iOS land I'm also scanning the Xamarin.Mac files here for simplicity.

### CODE REVIEW

File: msbuild/tests/Xamarin.iOS.Tasks.Tests/FrameworkListTest.cs [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+		}
+
+		List<AssemblyInfo> ScanFrameworkListXml (string frameworkListFile, bool isMac)
+		{
+			var assemblies = new List<AssemblyInfo> ();
+			var path = string.Empty;
+			if (isMac)
+				path = Path.GetFullPath (Path.Combine ("..", "..", "Xamarin.Mac.Tasks", frameworkListFile));
+			else
+				path = Path.GetFullPath (Path.Combine ("..", "..", "Xamarin.iOS.Tasks.Core", frameworkListFile));
```
Comment (spouliot) [inline msbuild/tests/Xamarin.iOS.Tasks.Tests/FrameworkListTest.cs]: DRY :)
```csharp
var path = Path.GetFullPath (Path.Combine ("..", "..", isMac ? "Xamarin.Mac.Tasks" : "Xamarin.iOS.Tasks.Core", frameworkListFile));
```

File: msbuild/tests/Xamarin.iOS.Tasks.Tests/FrameworkListTest.cs [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+		}
+
+		List<AssemblyInfo> ScanAssemblyDirectory (string frameworkName, bool isMac)
+		{
+			var assemblies = new List<AssemblyInfo> ();
+			var path = string.Empty;
+			if (isMac)
+				path = Path.GetFullPath (Path.Combine (Configuration.SdkRootXM, "lib", "mono", frameworkName));
+			else
+				path = Path.GetFullPath (Path.Combine (Configuration.MonoTouchRootDirectory, "lib", "mono", frameworkName));
```
Comment (spouliot) [inline msbuild/tests/Xamarin.iOS.Tasks.Tests/FrameworkListTest.cs]: same DRYness :)

File: msbuild/tests/Xamarin.iOS.Tasks.Tests/FrameworkListTest.cs [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+		public string PublicKeyToken;
+
+		public string Culture;
+
+		public ProcessorArchitecture ProcessorArchitecture = ProcessorArchitecture.MSIL;
+
+		public bool InGac;
+
+		public void Update (AssemblyName aname)
```
Comment (spouliot) [inline msbuild/tests/Xamarin.iOS.Tasks.Tests/FrameworkListTest.cs]: minor: that could have been the .ctor

File: msbuild/Xamarin.iOS.Tasks.Core/Xamarin.iOS-FrameworkList.xml.in [hunk 4]
```diff
@@ -2,4 +2,48 @@
 <FileList Name="Xamarin.iOS" TargetFrameworkDirectory="%TargetFrameworkDirectory%">
   <File AssemblyName="System.Buffers" Version="4.0.99.0" />
   <File AssemblyName="System.Memory" Version="4.0.99.0" />
```
Comment (spouliot) [inline msbuild/Xamarin.iOS.Tasks.Core/Xamarin.iOS-FrameworkList.xml.in]: why were those added ?
and why don't they include tokens and processor arch ? i.e. are those needed ?

File: msbuild/tests/Xamarin.iOS.Tasks.Tests/FrameworkListTest.cs [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+			else
+				path = Path.GetFullPath (Path.Combine (Configuration.MonoTouchRootDirectory, "lib", "mono", frameworkName));
+			foreach (var f in Directory.EnumerateFiles (path, "*.dll")) {
+				try {
+					var an = AssemblyName.GetAssemblyName (f);
+					var ainfo = new AssemblyInfo ();
+					ainfo.Update (an);
+					assemblies.Add (ainfo);
+				} catch (Exception ex) {
+					Console.WriteLine ("Error reading assembly '{0}' in framework '{1}':{2}{3}", f, frameworkName, Environment.NewLine, ex);
```
Comment (spouliot) [inline msbuild/tests/Xamarin.iOS.Tasks.Tests/FrameworkListTest.cs]: might be better to let the test fail in such case... otherwise we'll never realize there's a problem with one of the files

File: msbuild/Xamarin.iOS.Tasks.Core/Xamarin.iOS-FrameworkList.xml.in (hunk 4 above)
Comment (VincentDondain) [reply msbuild/Xamarin.iOS.Tasks.Core/Xamarin.iOS-FrameworkList.xml.in]: > why were those added ?

No clue, see https://github.com/xamarin/xamarin-macios/commit/6f2ebedb74fe8da250c8f53b61ffae2eb86fc711#diff-e801bb766cbaad95b50b1487b865f971?

> and why don't they include tokens and processor arch ?

The people adding that code didn't add that info and those assemblies are not in our "normal" assembly folder. To be fair I haven't looked for them too much, maybe they're in some other folder and I could update the test to report them properly...

> are those needed ?

Mikayla said so (:

Comment (VincentDondain) [reply msbuild/Xamarin.iOS.Tasks.Core/Xamarin.iOS-FrameworkList.xml.in]: @spouliot `/Library/Frameworks/Xamarin.iOS.framework/Versions/git/lib/mono/Xamarin.iOS/Facades`. I don't know if Xamarin.iOS/Mac should list everything in that Facades folder in `Xamarin.iOS-FrameworkList.xml.in`.

**Comment by chamons** [comment]
We could consider pushing this into a build step so we always generate these file instead of hard coding and then a test to tell us to fix them.

Given how likely they change though, it's likely fine.
